File: q152.cpp
class Solution
{
public:
    int maxProduct(vector<int> &nums)
    {
        if (nums.size() == 1)
            return nums[0];
        int currMax = nums[0];
        int currMin = nums[0];
        int prevMax = nums[0];
        int prevMin = nums[0];
        int ans = nums[0];
        for (int i = 1; i < nums.size(); i++)
        {
            currMax = max(max(prevMax * nums[i], prevMin * nums[i]), nums[i]);
            currMin = min(min(prevMax * nums[i], prevMin * nums[i]), nums[i]);
            ans = max(currMax, ans);
            prevMax = currMax;
            prevMin = currMin;
        }
        return ans;
    }
};

File: q152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        prev_min=nums[0]
        prev_max=nums[0]
        curr_min=nums[0]
        curr_max=nums[0]
        ans=nums[0]
        
        for x in nums[1:]:
            curr_min=min(prev_min*x,prev_max*x,x)
            curr_max=max(prev_min*x,prev_max*x,x)
            ans=max(curr_max,ans)
            prev_min=curr_min
            prev_max=curr_max
            
        return ans;
